pkgs/development/python-modules/pyarrow/default.nix

```diff
@@ -83,6 +83,10 @@ buildPythonPackage rec {
     export PYARROW_PARALLEL=$NIX_BUILD_CORES
   '';
 
+  # Builds the package twice: once for build_ext and once for bdist_wheel. This
+  # seems to be the only way to run the tests
+  setupPyBuildFlags = [ "--inplace" ];
```

Comment:
I'm not sure how to run the tests after installation of the binary wheel. I thought that was the default behavior in nixpkgs (`doInstallCheck` behaves this way IIUC and is how the Python infrastructure runs tests), but it seems like tests are run against the source, which without `--inplace` will not have the appropriate SOs in place to do anything.

Reply:
Did you try manually manipulating the `PYTHONPATH` before the check phase?

(assume you realize the `rm -r pyarrow/!(tests)` that you removed below was originally to prevent the tests running against the source dir)

Reply:
If I revert this, it fixes the `pythonImportsCheckPhase` on darwin, but breaks the `pytestCheckPhase`:
```
pytestCheckPhase
Executing pytestCheckPhase
============================= test session starts ==============================
platform darwin -- Python 3.10.5, pytest-7.1.2, pluggy-1.0.0
rootdir: /private/tmp/nix-build-python3.10-pyarrow-9.0.0.drv-1/apache-arrow-9.0.0/python, configfile: setup.cfg
plugins: hypothesis-6.50.1, lazy-fixture-0.6.3
collected 0 items / 1 error                                                    

==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
/nix/store/rs23kn5r3qm96cq26b6bzy2w7rs5r39d-python3-3.10.5/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1050: in _gcd_import
    ???
<frozen importlib._bootstrap>:1027: in _find_and_load
    ???
<frozen importlib._bootstrap>:1006: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:688: in _load_unlocked
    ???
/nix/store/0fq7xcm395b20lwbrcsgnfsijfg420df-python3.10-pytest-7.1.2/lib/python3.10/site-packages/_pytest/assertion/rewrite.py:168: in exec_module
    exec(co, module.__dict__)
pyarrow/tests/conftest.py:26: in <module>
    from ..conftest import groups, defaults
E   ImportError: attempted relative import beyond top-level package
=========================== short test summary info ============================
ERROR  - ImportError: attempted relative import beyond top-level package
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.86s ===============================
```

Reply:
As I understand, pytest is loading https://github.com/apache/arrow/blob/apache-arrow-9.0.0/python/pyarrow/tests/conftest.py, which relatively imports https://github.com/apache/arrow/blob/apache-arrow-9.0.0/python/pyarrow/conftest.py, but the sources were removed by the `preCheck`.

Reply:
This fixes the import error:

``` diff
diff --git i/pkgs/development/python-modules/pyarrow/default.nix w/pkgs/development/python-modules/pyarrow/default.nix
index 53dde0cd0f6..fd20e0a5ede 100644
--- i/pkgs/development/python-modules/pyarrow/default.nix
+++ w/pkgs/development/python-modules/pyarrow/default.nix
@@ -111,7 +111,7 @@ buildPythonPackage rec {
 
   preCheck = ''
     shopt -s extglob
-    rm -r pyarrow/!(tests)
+    rm -r pyarrow/!(__init__.py|conftest.py|tests)
   '' + lib.optionalString stdenv.isDarwin  ''
     # OSError: [Errno 24] Too many open files
     ulimit -n 1024
```

But now there is a new error:

```
python3.10-pyarrow> ============================= test session starts ==============================
python3.10-pyarrow> platform darwin -- Python 3.10.5, pytest-7.1.2, pluggy-1.0.0
python3.10-pyarrow> rootdir: /private/tmp/nix-build-python3.10-pyarrow-9.0.0.drv-0/apache-arrow-9.0.0/python, configfile: setup.cfg
python3.10-pyarrow> plugins: hypothesis-6.50.1, lazy-fixture-0.6.3
python3.10-pyarrow> collected 0 items / 1 error
python3.10-pyarrow> ==================================== ERRORS ====================================
python3.10-pyarrow> ________________________ ERROR collecting test session _________________________
python3.10-pyarrow> pyarrow/__init__.py:40: in <module>
python3.10-pyarrow>     from ._generated_version import version as __version__
python3.10-pyarrow> E   ModuleNotFoundError: No module named 'pyarrow._generated_version'
python3.10-pyarrow> During handling of the above exception, another exception occurred:
python3.10-pyarrow> /nix/store/rs23kn5r3qm96cq26b6bzy2w7rs5r39d-python3-3.10.5/lib/python3.10/importlib/__init__.py:126: in import_module
python3.10-pyarrow>     return _bootstrap._gcd_import(name[level:], package, level)
python3.10-pyarrow> <frozen importlib._bootstrap>:1050: in _gcd_import
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:1027: in _find_and_load
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:992: in _find_and_load_unlocked
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:241: in _call_with_frames_removed
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:1050: in _gcd_import
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:1027: in _find_and_load
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:1006: in _find_and_load_unlocked
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:688: in _load_unlocked
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap_external>:883: in exec_module
python3.10-pyarrow>     ???
python3.10-pyarrow> <frozen importlib._bootstrap>:241: in _call_with_frames_removed
python3.10-pyarrow>     ???
python3.10-pyarrow> pyarrow/__init__.py:56: in <module>
python3.10-pyarrow>     __version__ = setuptools_scm.get_version('../',
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/__init__.py:147: in get_version
python3.10-pyarrow>     maybe_version = _get_version(config)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/__init__.py:154: in _get_version
python3.10-pyarrow>     parsed_version = _do_parse(config)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/__init__.py:88: in _do_parse
python3.10-pyarrow>     parse_result = _call_entrypoint_fn(config.absolute_root, config, config.parse)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/_entrypoints.py:49: in _call_entrypoint_fn
python3.10-pyarrow>     return fn(root)
python3.10-pyarrow> pyarrow/__init__.py:55: in parse_git
python3.10-pyarrow>     return parse(root, **kwargs)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/git.py:179: in parse
python3.10-pyarrow>     wd = get_working_directory(config)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/git.py:164: in get_working_directory
python3.10-pyarrow>     return GitWorkdir.from_potential_worktree(config.absolute_root)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/git.py:54: in from_potential_worktree
python3.10-pyarrow>     require_command(cls.COMMAND)
python3.10-pyarrow> /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/utils.py:171: in require_command
python3.10-pyarrow>     raise OSError("%r was not found" % name)
python3.10-pyarrow> E   OSError: 'git' was not found
python3.10-pyarrow> =============================== warnings summary ===============================
python3.10-pyarrow> ../../../../../nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/__init__.py:88
python3.10-pyarrow>   /nix/store/h84rrja5lzg0f8fba2vbbab0kh3ypvhx-python3.10-setuptools-scm-7.0.5/lib/python3.10/site-packages/setuptools_scm/__init__.py:88: DeprecationWarning: parse function pyarrow.parse_git are required to provide a named argument 'config', setuptools_scm>=8.0 will remove support.
python3.10-pyarrow>     parse_result = _call_entrypoint_fn(config.absolute_root, config, config.parse)
python3.10-pyarrow> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
python3.10-pyarrow> =========================== short test summary info ============================
python3.10-pyarrow> ERROR  - OSError: 'git' was not found
python3.10-pyarrow> !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
python3.10-pyarrow> ========================= 1 warning, 1 error in 0.24s ==========================
```

Looks like `git` is expected, but probably a patch is better. I'll try to figure out.

Reply:
Sorry, I messed up on the previous comment and I deleted it.

This patch fixes the import error:

``` diff
diff --git i/pkgs/development/python-modules/pyarrow/default.nix w/pkgs/development/python-modules/pyarrow/default.nix
index 53dde0cd0f6..eef0057d727 100644
--- i/pkgs/development/python-modules/pyarrow/default.nix
+++ w/pkgs/development/python-modules/pyarrow/default.nix
@@ -111,7 +111,9 @@ buildPythonPackage rec {
 
   preCheck = ''
     shopt -s extglob
-    rm -r pyarrow/!(tests)
+    rm -r pyarrow/!(conftest.py|tests)
+    mv pyarrow/conftest.py pyarrow/tests/parent_conftest.py
+    substituteInPlace pyarrow/tests/conftest.py --replace ..conftest .parent_conftest
   '' + lib.optionalString stdenv.isDarwin  ''
     # OSError: [Errno 24] Too many open files
     ulimit -n 1024
```

We need `pyarrow/conftest.py` but cannot be in that location, otherwise `pyarrow/__init__.py` would also be needed, which would make the empty `pyarrow` directory a package instead of using the installed package. So as a workaround, I move it to `pyarrow/tests/parent_conftest.py` (or any other name).

But the tests fail:

```
python3.10-pyarrow> ============================= test session starts ==============================
python3.10-pyarrow> platform darwin -- Python 3.10.5, pytest-7.1.2, pluggy-1.0.0
python3.10-pyarrow> rootdir: /private/tmp/nix-build-python3.10-pyarrow-9.0.0.drv-0/apache-arrow-9.0.0/python, configfile: setup.cfg
python3.10-pyarrow> plugins: hypothesis-6.50.1, lazy-fixture-0.6.3
python3.10-pyarrow> collected 4302 items / 13 deselected / 3 skipped / 4289 selected
python3.10-pyarrow> pyarrow/tests/test_adhoc_memory_leak.py s                                [  0%]
python3.10-pyarrow> pyarrow/tests/test_array.py ........................s................... [  1%]
python3.10-pyarrow> ........................................................................ [  2%]
python3.10-pyarrow> .....s.................................................................. [  4%]
python3.10-pyarrow> ..............ss.............                                            [  5%]
python3.10-pyarrow> pyarrow/tests/test_builder.py ....                                       [  5%]
python3.10-pyarrow> pyarrow/tests/test_cffi.py ...........                                   [  5%]
python3.10-pyarrow> pyarrow/tests/test_compute.py .......................................... [  6%]
python3.10-pyarrow> ........................................................................ [  8%]
python3.10-pyarrow> ........................................................................ [  9%]
python3.10-pyarrow> .........................................                                [ 10%]
python3.10-pyarrow> pyarrow/tests/test_convert_builtin.py .................................. [ 11%]
python3.10-pyarrow> ........................................................................ [ 13%]
python3.10-pyarrow> ........................................................................ [ 14%]
python3.10-pyarrow> .................................x...................................... [ 16%]
python3.10-pyarrow> ................................................x....................... [ 18%]
python3.10-pyarrow> ssssss.................................................................. [ 19%]
python3.10-pyarrow> ......................................sssssssss                          [ 21%]
python3.10-pyarrow> pyarrow/tests/test_csv.py .............................................. [ 22%]
python3.10-pyarrow> ..................................................................s...   [ 23%]
python3.10-pyarrow> pyarrow/tests/test_cython.py ..                                          [ 23%]
python3.10-pyarrow> pyarrow/tests/test_dataset.py .......................................... [ 24%]
python3.10-pyarrow> ........................................................................ [ 26%]
python3.10-pyarrow> ............................................EEEE....................ssss [ 28%]
python3.10-pyarrow> ...s...s............................................................EE.. [ 29%]
python3.10-pyarrow> ....                                                                     [ 29%]
python3.10-pyarrow> pyarrow/tests/test_exec_plan.py ........................................ [ 30%]
python3.10-pyarrow>                                                                          [ 30%]
python3.10-pyarrow> pyarrow/tests/test_extension_type.py ................................... [ 31%]
python3.10-pyarrow> ......                                                                   [ 31%]
python3.10-pyarrow> pyarrow/tests/test_feather.py .............................ss........... [ 32%]
python3.10-pyarrow> .....................ss....x........s.....s.                             [ 33%]
python3.10-pyarrow> pyarrow/tests/test_filesystem.py .....                                   [ 33%]
python3.10-pyarrow> pyarrow/tests/test_flight.py .FFsFFFFFFFFFFFF.sFFFFFsFFFFFFFFF..sFFFFFFF [ 34%]
python3.10-pyarrow> ss.FFFFFFFFFFFFFFsF                                                      [ 35%]
python3.10-pyarrow> pyarrow/tests/test_fs.py E.......Essx.x..E...Essx.x..E....Ess.....E...Es [ 36%]
python3.10-pyarrow> s.....E...Ess.....E...Ess.....E...Ess.....E...Ess.....E...Ess.....E....E [ 38%]
python3.10-pyarrow> ss.....E...Ess.....E...Ess.....E...Ess.....E............EEEEssssssss.... [ 39%]
python3.10-pyarrow> ................EEEE...Ess.....E...Ess.....E............EEEEssssssss.... [ 41%]
python3.10-pyarrow> ................EEEE............EEEEssssssss....................EEEE...E [ 43%]
python3.10-pyarrow> ss.....E....s.s...............Es...........E.                            [ 44%]
python3.10-pyarrow> pyarrow/tests/test_gandiva.py ssssssssssss                               [ 44%]
python3.10-pyarrow> pyarrow/tests/test_gdb.py ssssssssssssssssssssssss                       [ 45%]
python3.10-pyarrow> pyarrow/tests/test_hdfs.py ssssssssssssssssssssssss                      [ 45%]
python3.10-pyarrow> pyarrow/tests/test_io.py ........s...........................s.....s.... [ 46%]
python3.10-pyarrow> ...........s...s..................s....s......s....xs................... [ 48%]
python3.10-pyarrow> .............                                                            [ 48%]
python3.10-pyarrow> pyarrow/tests/test_ipc.py .............................................. [ 49%]
python3.10-pyarrow> ............                                                             [ 50%]
python3.10-pyarrow> pyarrow/tests/test_json.py ..........................                    [ 50%]
python3.10-pyarrow> pyarrow/tests/test_memory.py ........................                    [ 51%]
python3.10-pyarrow> pyarrow/tests/test_misc.py .s........................................... [ 52%]
python3.10-pyarrow> ....................................................                     [ 53%]
python3.10-pyarrow> pyarrow/tests/test_orc.py sssssssssssss                                  [ 53%]
python3.10-pyarrow> pyarrow/tests/test_pandas.py ........................................... [ 54%]
python3.10-pyarrow> ....................................s............................sss.... [ 56%]
python3.10-pyarrow> .................................................s.........s............ [ 58%]
python3.10-pyarrow> ................................s....................................... [ 59%]
python3.10-pyarrow> ..............................................                           [ 60%]
python3.10-pyarrow> pyarrow/tests/test_plasma.py sssssssssssssssssssssssssssssssss           [ 61%]
python3.10-pyarrow> pyarrow/tests/test_plasma_tf_op.py s                                     [ 61%]
python3.10-pyarrow> pyarrow/tests/test_scalars.py .......................................... [ 62%]
python3.10-pyarrow> ....s....s....................                                           [ 63%]
python3.10-pyarrow> pyarrow/tests/test_schema.py ................................            [ 64%]
python3.10-pyarrow> pyarrow/tests/test_serialization.py ...........ss....................... [ 64%]
python3.10-pyarrow> ........................................................................ [ 66%]
python3.10-pyarrow> ........................................................................ [ 68%]
python3.10-pyarrow> ..............s......................................................... [ 69%]
python3.10-pyarrow> ........................................................................ [ 71%]
python3.10-pyarrow> ........................................................................ [ 73%]
python3.10-pyarrow> ........................................................................ [ 74%]
python3.10-pyarrow> ........................................................................ [ 76%]
python3.10-pyarrow> ........................................................................ [ 78%]
python3.10-pyarrow> ........................................................................ [ 80%]
python3.10-pyarrow> ............................................s..............              [ 81%]
python3.10-pyarrow> pyarrow/tests/test_serialization_deprecated.py ..                        [ 81%]
python3.10-pyarrow> pyarrow/tests/test_sparse_tensor.py .................................... [ 82%]
python3.10-pyarrow> ........................................................................ [ 83%]
python3.10-pyarrow> ..............................................sssssssssss                [ 85%]
python3.10-pyarrow> pyarrow/tests/test_strategies.py ssssssss                                [ 85%]
python3.10-pyarrow> pyarrow/tests/test_substrait.py sss                                      [ 85%]
python3.10-pyarrow> pyarrow/tests/test_table.py .......................s.................... [ 86%]
python3.10-pyarrow> .................................................................        [ 88%]
python3.10-pyarrow> pyarrow/tests/test_tensor.py .....................                       [ 88%]
python3.10-pyarrow> pyarrow/tests/test_types.py ....................s....................... [ 89%]
python3.10-pyarrow> ..............................ss                                         [ 90%]
python3.10-pyarrow> pyarrow/tests/test_udf.py .................                              [ 90%]
python3.10-pyarrow> pyarrow/tests/test_util.py s                                             [ 90%]
python3.10-pyarrow> pyarrow/tests/parquet/test_basic.py .................s.................. [ 91%]
python3.10-pyarrow> ......................s......................................            [ 93%]
python3.10-pyarrow> pyarrow/tests/parquet/test_compliant_nested_type.py ........             [ 93%]
python3.10-pyarrow> pyarrow/tests/parquet/test_data_types.py ..............X................ [ 93%]
python3.10-pyarrow> .sssssss.ss                                                              [ 94%]
python3.10-pyarrow> pyarrow/tests/parquet/test_dataset.py ....x..........xx.......x...x.EEEE [ 94%]
python3.10-pyarrow> ..EEEE..........................x.x...........EEEEEE.s.................  [ 96%]
python3.10-pyarrow> pyarrow/tests/parquet/test_datetime.py ...................               [ 97%]
python3.10-pyarrow> pyarrow/tests/parquet/test_encryption.py ssssssssssssssss                [ 97%]
python3.10-pyarrow> pyarrow/tests/parquet/test_metadata.py .........................s.       [ 98%]
python3.10-pyarrow> pyarrow/tests/parquet/test_pandas.py .......s........................... [ 98%]
python3.10-pyarrow> ..............                                                           [ 99%]
python3.10-pyarrow> pyarrow/tests/parquet/test_parquet_file.py ..............                [ 99%]
python3.10-pyarrow> pyarrow/tests/parquet/test_parquet_writer.py .............EEE...         [100%]
```

Reply:
The tests pass if I disable the darwin sandbox.